File: packages/mylife-money/src/business/views/total-by-month.ts
import { StoreMaterializedView, StoreCollection, StorePartitionIndex, getStoreCollection, getMetadataEntity } from 'mylife-tools-server';
import { dateToMonth, roundCurrency, monthRange } from './tools';

export class TotalByMonth extends StoreMaterializedView {
  private readonly collection: StoreCollection;

  constructor() {
    super('total-by-month', getMetadataEntity('report-total-by-month'));
    this.collection = getStoreCollection('operations');
    this.collection.on('change', this.onCollectionChange);

    this.init();
  }

  close() {
    this.collection.off('change', this.onCollectionChange);
    this._reset();
  }

  private init() {
    for (const month of getMonthList()) {
      this.computeBucket(month);
    }
  }

  private readonly onCollectionChange = ({ before, after, type }) => {
    const months = getMonthList();

    switch(type) {
      case 'create': {
        const month = dateToMonth(after.date);

        if (months.has(month)) {
          this.computeBucket(month);
        }

        break;
      }
  
      case 'update': {
        const month1 = dateToMonth(before.date);
        const month2 = dateToMonth(after.date);

        // should be useless since amount cannot change
        if (months.has(month1)) {
          this.computeBucket(month1);
        }

        if (month1 !== month2 && months.has(month2)) {
          // should not happen since date cannot change
          this.computeBucket(month2);
        }

        break;
      }
  
      case 'remove': {
        // should not happen since operations cannot be removed
        const month = dateToMonth(before.date);

        if (months.has(month)) {
          this.computeBucket(month);
        }

        break;
      }
  
      default:
        throw new Error(`Unsupported event type: '${type}'`);
    }

    this.cleanOldBuckets(months);
  };

  private computeBucket(month: string) {
    const index = this.collection.getIndex('partition-by-month') as StorePartitionIndex;
    const partition = index.findPartition(month) || new Set();

    const object = {
      _id: month,
      month,
      count: 0,
      sumDebit: 0,
      sumCredit: 0,
      balance: 0
    };

    for (const operation of partition) {
      ++object.count;
      operation.amount < 0 ? (object.sumDebit += -operation.amount) : (object.sumCredit += operation.amount);
      object.balance += operation.amount;
    }

    object.sumDebit = roundCurrency(object.sumDebit);
    object.sumCredit = roundCurrency(object.sumCredit);
    object.balance = roundCurrency(object.balance);

    this._set(this.entity.newObject(object));
  }

  private cleanOldBuckets(months: Set<string>) {
    for (const object of this.list()) {
      if (!months.has(object._id)) {
        this._delete(object._id);
      }
    }
  }
}

function getMonthList() {
  const now = new Date();
  const minDate = new Date(now.getFullYear() - 1, 0, 1);
  return new Set(monthRange(minDate, now));
}

File: go/mylife-money/pkg/business/views/total_by_month.go
package views

import (
	"fmt"
	"mylife-money/pkg/entities"
	"mylife-tools-server/services/store"
	"time"
)

type totalByMonth struct {
	operations     store.ICollection[*entities.Operation]
	container      *store.Container[*entities.ReportTotalByMonth]
	changeCallback func(*store.Event[*entities.Operation])
}

var _ store.IView[*entities.ReportTotalByMonth] = (*totalByMonth)(nil)

func (view *totalByMonth) AddListener(callback *func(event *store.Event[*entities.ReportTotalByMonth])) {
	view.container.AddListener(callback)
}

func (view *totalByMonth) RemoveListener(callback *func(event *store.Event[*entities.ReportTotalByMonth])) {
	view.container.RemoveListener(callback)
}

func (view *totalByMonth) Name() string {
	return view.container.Name()
}

func (view *totalByMonth) Find(id string) (*entities.ReportTotalByMonth, bool) {
	return view.container.Find(id)
}

func (view *totalByMonth) Get(id string) (*entities.ReportTotalByMonth, error) {
	return view.container.Get(id)
}

func (view *totalByMonth) List() []*entities.ReportTotalByMonth {
	return view.container.List()
}

func (view *totalByMonth) Size() int {
	return view.container.Size()
}

func (view *totalByMonth) Filter(predicate func(obj *entities.ReportTotalByMonth) bool) []*entities.ReportTotalByMonth {
	return view.container.Filter(predicate)
}

func (view *totalByMonth) Exists(predicate func(obj *entities.ReportTotalByMonth) bool) bool {
	return view.container.Exists(predicate)
}

func (view *totalByMonth) Close() {
	view.operations.RemoveListener(&view.changeCallback)
	view.container.Reset()
}

func (view *totalByMonth) Refresh() {
	view.init()
}

func NewTotalByMonth() (store.IView[*entities.ReportTotalByMonth], error) {
	operations, err := store.GetCollection[*entities.Operation]("operations")
	if err != nil {
		return nil, err
	}

	container := store.NewContainer[*entities.ReportTotalByMonth]("total-by-month")

	view := &totalByMonth{
		operations: operations,
		container:  container,
	}

	view.changeCallback = view.onCollectionChange
	view.operations.AddListener(&view.changeCallback)

	view.init()

	return view, nil
}

func (view *totalByMonth) init() {
	months := getMonthList()
	for month := range months {
		view.computeBucket(month)
	}
}

func (view *totalByMonth) onCollectionChange(event *store.Event[*entities.Operation]) {
	months := getMonthList()

	switch event.Type() {
	case store.Create:
		month := DateToMonth(event.After().Date())

		if _, ok := months[month]; ok {
			view.computeBucket(month)
		}

	case store.Update:
		month1 := DateToMonth(event.Before().Date())
		month2 := DateToMonth(event.After().Date())

		// should be useless since amount cannot change
		if _, ok := months[month1]; ok {
			view.computeBucket(month1)
		}

		if _, ok := months[month2]; month1 != month2 && ok {
			// should not happen since date cannot change
			view.computeBucket(month2)
		}

	case store.Remove:
		// should not happen since operations cannot be removed
		month := DateToMonth(event.Before().Date())

		if _, ok := months[month]; ok {
			view.computeBucket(month)
		}

	default:
		panic(fmt.Sprintf("Unsupported event type: '%v'", event.Type()))
	}

	view.cleanOldBuckets(months)
}

func (view *totalByMonth) computeBucket(month string) {
	untypedIndex, err := view.operations.GetIndex("partition-by-month")
	if err != nil {
		panic(fmt.Sprintf("Failed to get index: %v", err))
	}

	index := untypedIndex.(store.IPartitionIndex[*entities.Operation, string])

	partition := index.FindPartition(month)
	if partition == nil {
		partition = make(map[string]*entities.Operation)
	}

	object := &entities.ReportTotalByMonthValues{
		Id:        month,
		Month:     month,
		Count:     0,
		SumDebit:  0,
		SumCredit: 0,
		Balance:   0,
	}

	for _, operation := range partition {
		object.Count++

		if operation.Amount() < 0 {
			object.SumDebit += -operation.Amount()
		} else {
			object.SumCredit += operation.Amount()
		}

		object.Balance += operation.Amount()
	}

	object.SumDebit = roundCurrency(object.SumDebit)
	object.SumCredit = roundCurrency(object.SumCredit)
	object.Balance = roundCurrency(object.Balance)

	view.container.Set(entities.NewReportTotalByMonth(object))
}

func (view *totalByMonth) cleanOldBuckets(months map[string]struct{}) {
	for _, object := range view.List() {
		id := object.Id()

		if _, ok := months[id]; !ok {
			view.container.Delete(id)
		}
	}
}

func getMonthList() map[string]struct{} {
	now := time.Now()
	minDate := time.Date(now.Year(), 1, 1, 0, 0, 0, 0, time.UTC)
	array := monthRange(minDate, now)

	set := make(map[string]struct{})

	for _, month := range array {
		set[month] = struct{}{}
	}

	return set
}
